### PR TITLE
feature: Masked() and MaskedToBuf() — apply bitmask to keys and merge colliding containers

### PR DESCRIPTION
**Pull request by aliszka**
## Motivation

  Bitmap keys encode dimensions in the upper bits. Callers need to project a bitmap onto a sub-dimension by masking out the upper bits and merging containers whose keys collide under the mask — without materialising an intermediate bitmap per dimension.

  ## Approach

  `Masked(mask uint64)` iterates over all source containers, applies the mask to each key, and OR-merges containers that collapse to the same masked key into the result. The lowest 16 bits of the mask are forced to zero since keys never use them.

  Two optimisations in the merge loop keep allocations minimal:
  - Key space is pre-sized upfront (`expandConditionally(an, 0)`) so the loop never shifts container data to make room for new keys.
  - When an inline OR fails (container grew beyond its current allocation) and the old container sits at the tail of `b.data`, the slice is trimmed and regrown in place, avoiding a dead container fragment.

  `MaskedToBuf(mask, buf)` is the same operation but writes into a caller-supplied byte slice via `NewBitmapToBuf`, avoiding heap allocation when the caller can reuse a pre-allocated buffer.

  ## Key areas for review

  - `bitmap_opt.go:maskedInto` — the trim-and-regrow path (`b.data = b.data[:boff]`) is only safe when `boff+len(bc) == len(b.data)`; verify the condition is tight
  - `bitmap_opt.go:maskedInto` — `expandConditionally(an, 0)` pre-sizes for the worst case (no key collisions); with heavy collisions the key space is over-allocated, but this is bounded by the source key count

  ## Risks and mitigations

  - **Trim-and-regrow leaves dead space when container is not at tail**: accepted trade-off; sequential key iteration almost always produces containers in append order, so the tail condition hits in the common case
  - **Mask low-bit truncation is silent**: callers passing a mask with low 16 bits set get them silently cleared — documented on the method

  ## Testing

  Tests cover nil input, no key overlap, key collision with OR merge (including multi-key collapse under zero mask), mask boundary conditions (zero mask, identity mask, low-16-bits ignored), and the buffer-swap logic in the OR accumulation loop for non-overlapping arrays, overlapping arrays, array-to-bitmap conversion, and three-container chains.

### CODE REVIEW

File: bitmap_opt.go [hunk 1]
```diff
@@ -1176,3 +1176,72 @@ func (ra *Bitmap) expandConditionally(newKeys int, sizeContainers int) {
...
+	if ra == nil {
+		return b
+	}
+	an := ra.keys.numKeys()
+	if an == 0 {
+		return b
+	}
+
+	// Ensure the mask has its lowest 16 bits unset, since keys always do.
+	mask &= 0xFFFFFFFFFFFF0000
```
Comment (amourao) [inline bitmap_opt.go:1201]: **suggestion:** instead of documentation only, we could err out here if the mask doesn't have the lowest 16 bits unset (and make MaskedToBuf also return an error

File: bitmap_opt.go [hunk 2]
```diff
@@ -1176,3 +1176,72 @@ func (ra *Bitmap) expandConditionally(newKeys int, sizeContainers int) {
...
+		}
+
+		maskedKey := ak & mask
+
+		boff, has := b.keys.getValue(maskedKey)
+		if !has {
+			// First container for this masked key — copy it directly.
+			b.expandConditionally(0, len(ac))
+			boff = b.newContainerNoClr(uint16(len(ac)))
+			copy(b.data[boff:], ac)
```
Comment (amourao) [inline bitmap_opt.go:1224]: **question:** this may be an incredibly dumb question, but would it be possible to not copy the full container memory if a single container matches for the key and we know the og bitmap isn't mutated? This is def an optimization that may not be worth it, but it could work in some scenario? I apologize in advance 😅

File: bitmap.go [hunk 3]
```diff
@@ -100,6 +100,28 @@ func NewBitmapWith(numKeys int) *Bitmap {
...
+// unsafe []uint16 view of the same memory.
+func NewBitmapToBuf(buf []byte) *Bitmap {
+	if len(buf)%2 != 0 {
+		buf = buf[:len(buf)-1]
+	}
+	numKeys := 2
+	keysLen := calcInitialKeysLen(numKeys)
+	minBytes := (keysLen + minContainerSize) * 2
+	if len(buf) < minBytes {
+		panic(fmt.Sprintf("buf too small: need at least %d bytes, got %d", minBytes, len(buf)))
```
Comment (amourao) [inline bitmap.go]: **question:** would returning an error instead be better? I can see this crashing Weaviate from some misunderstanding, being so far down the stack

File: bitmap_test.go [hunk 4]
```diff
@@ -386,6 +515,312 @@ func TestSetSorted(t *testing.T) {
...
+
+		bufSize := 1 << 20 // 1MB
+		result := bm.MaskedToBuf(math.MaxUint64, make([]byte, bufSize))
+
+		require.Equal(t, 50, result.GetCardinality())
+		require.Greater(t, result.LenInBytes(), 0)
+		require.Equal(t, bufSize, result.capInBytes(), "capacity should not change")
+	})
+}
+
```
Comment (amourao) [inline bitmap_test.go:816]: **question:** does any of the tests explicitly tests two colliding Array containers forces an upgrade to a Bitmap container at the tail end of b.data?

**Review by amourao — APPROVED** [review]
LGTM, I have a couple of comments, but they may miss some context as I haven't checked where this code is called by Weaviate

File: bitmap.go (hunk 3 above)
Comment (aliszka) [reply bitmap.go]: errors are very rarely returned in sroar library as handling them can be painful later.
usually panics are used, but they serve as a guards and should be revealed only in the tests, impossible to be triggered by a user. not the case here - as user is responsible for providing buffer of valid size.

i think safest approach would be to ignore buffer that is too small and silently create new one big enough to contain empty bitmap.

File: bitmap_opt.go (hunk 1 above)
Comment (aliszka) [reply bitmap_opt.go:1201]: while i understand your point, i wanted to deliberately avoid returning an error to simplify method usage. And masks that weaviate will be using will not contain lower bits.

File: bitmap_opt.go (hunk 2 above)
Comment (aliszka) [reply bitmap_opt.go:1224]: not sure i understood the question, so the reply may not be valid.

`Masked` methods do not mutate input bitmap, they always create new one, so copying is necessary to populate masked bitmap. Here container is copied entirely as relevant maskedKey is not yet present in the destination bitmap. Later if same maskedKey will be encountered in src bitmap again, copied container and src container will be OR-merged in dst bitmap (inline if possible)

I will try to figure out if there is a performant way to apply mask by mutating existing bitmap (without creating new one), but that is a task for the future

File: bitmap_test.go (hunk 4 above)
Comment (aliszka) [reply bitmap_test.go:816]: there is not. great catch! will add some

File: bitmap.go (hunk 3 above)
Comment (aliszka) [reply bitmap.go]: changed, to silently replace the buffer with bigger one if the one provided is too small

File: bitmap_test.go (hunk 4 above)
Comment (aliszka) [reply bitmap_test.go:816]: added 2 more tests:
- array + array => bigger array
- array + array => bitmap
